zulip_bots/zulip_bots/test_lib.py

```diff
@@ -70,29 +71,58 @@ def check_expected_responses(self, expectations, expected_method='send_reply',
+                                          state_handler=state_handler)
 
-    def call_request(self, message, expected_method, response, state_handler):
-        # type: (Dict[str, Any], str, Dict[str, Any], Optional[StateHandler]) -> None
+    def call_request(self, message, *responses, **kwargs):
+        # type: (Dict[str, Any], *Tuple[Dict[str, Any], str], **Any) -> None
+        state_handler = kwargs.pop('state_handler', None)
         if state_handler is None:
             state_handler = StateHandler()
-        # Send message to the concerned bot
-        self.message_handler.handle_message(message, self.MockClass(), state_handler)
 
-        # Check if the bot is sending a message via `send_message` function.
-        # Where response is a dictionary here.
+        # Mock BotHandler; get instance
+        bot_handler = self.MockClass()
         instance = self.MockClass.return_value
-        if expected_method == "send_message":
-            instance.send_message.assert_called_with(response)
-        else:
-            instance.send_reply.assert_called_with(message, response['content'])
+
+        # Send message to the bot
+        try:
+            self.message_handler.handle_message(message, bot_handler, state_handler)
+        except KeyError as key_err:
+            raise Exception("Message tested likely required key {}.".format(key_err))
+
+        # Determine which messaging functions are expected
+        send_messages = [call(r[0]) for r in responses if r[1] == 'send_message']
+        send_replies = [call(message, r[0]['content']) for r in responses if r[1] == 'send_reply']
+
+        # Test that call were matching in quantity, and then in details
+        fail_template = "\nMESSAGE:\n{}\nACTUAL CALLS:\n{}\nEXPECTED:\n{}\n"
+        functions_to_test = (('send_message', instance.send_message, send_messages),
+                             ('send_reply', instance.send_reply, send_replies))
+        for version, actual, expected in functions_to_test:
+            err = None
+            try:
+                assert(len(expected) == actual.call_count)
+            except AssertionError:
+                err = ("Numbers of {} called do not match those expected ({} calls, {} expected)" +
+                       fail_template).format(version, actual.call_count, len(expected),
+                                             message, actual.call_args_list, expected)
+            else:
+                try:
+                    if len(expected) > 0:
+                        actual.assert_has_calls(expected)  # any_order = True FIXME
+                except AssertionError:
+                    err = ("Calls to {} do not match those expected" +
+                           fail_template).format(version,
+                                                 message, actual.call_args_list, expected)
+            if err is not None:
+                raise AssertionError(err)
```

Comment:
I tried something like this, with simply `print(err); raise` after each `err=...` assignment, but this shifts the output to the start of the test output. For example, the (current) error message in master in the weather-bot without it:
```
AssertionError: Calls to send_reply do not match those expected
MESSAGE:
{'content': 'United Kingdom'}
ACTUAL CALLS:
[call({'content': 'United Kingdom'}, 'Weather in London, GB:\n58.73 F / 14.85 C\nShower Rain')]
EXPECTED:
[call({'content': 'United Kingdom'}, 'Weather in London, GB:\n58.33 F / 14.85 C\nShower Rain')]
```
is then printed right near the start (away from the traceback where you'd see it), with the 'final' text then only:
```
AssertionError: Calls not found.
Expected: [call({'content': 'United Kingdom'}, 'Weather in London, GB:\n58.33 F / 14.85 C\nShower Rain')]
Actual: [call({'content': 'United Kingdom'}, 'Weather in London, GB:\n58.73 F / 14.85 C\nShower Rain')]
```
(I actually thought the text was just not being printed for some reason)

With python3 the blocks can be of the form of the following, and works more like you want:
```
try:
    [...]
except AssertionError as e:
    raise AssertionError("...") from e
```
...but apparently this is new in python3, so isn't an option here?

The rest of the traceback is the same if I re-raise; is it the case of accidentally catching other assert() failures in code that you're wanting to avoid? If so, would it be simpler to just switch to using `if`, at least for the first test?